src/hooks/use-get-articles-query-params.ts

```diff
@@ -0,0 +1,35 @@
+import { useAppSelector } from '@inet/app/hooks'
+import {
+  ArticleCreatedWithin,
+  IGetArticlesQuery,
+} from '@inet/services/types/articles/i-get-articles.query'
+import { Pagination } from '@inet/services/types/shared/pagination'
+import { useSearchParams } from 'react-router-dom'
+
+const useGetArticlesQueryParams = (
```

Comment:
報告史蒂芬哥
今天網路上又查了幾篇關於useGetQuery到底應該把response存進store裡還是params存進store裡面的問題
有一篇堆疊溢出是說在任何需要那個response的地方盡情的呼叫useGetQuery
還有人提到官網的這段話 

> However, when you use RTK Query, there is a mindset shift that happens. We're no longer thinking about "managing state" per se. Instead, we now think about "managing cached data". Rather than trying to write reducers ourselves, we're now going to focus on defining "where is this data coming from?", "how should this update be sent?", "when should this cached data be re-fetched?", and "how should the cached data be updated?". How that data gets fetched, stored, and retrieved becomes implementation details we no longer have to worry about.

雖然我是有看沒有懂

而chatGPT是說用相同的params去呼叫就好
所以我還是試著把呼叫這個query會用到的params通通存進store裡 (除了史蒂芬哥您的偉大pagination hook)
之後看看用起來如何再向您報告

Reply:
報告patrick哥
RTK Query 的 response 是存在記憶體裡面的  在 cache 過期前他都會直接回傳那個存起來的數值
如果有些元件要共用某個 api 的 response  傳統的作法是：

1. 在盡量靠近上層的元件呼叫api
2. 把 response 存進 store
3. 在需要那個 response 元件中取出 store 的值來用

所以實作的時候我們就必須多加小心  確定只有一個地方呼叫那個大家都需要的 API
才不會做了不必要的呼叫  或是後面蓋前面

---

但是使用 RTK Query 的時候我們就不用管那麼多
他的每個 endpoint 都有自己的狀態
而且這些狀態只有一個實體  可以看成是 singleton
可以把他的 store 想像成這樣

```ts
const store = {
  getSomethingQuery: {
    isLoading: false,
    data: undefined,
    previousArgs: undefined,
    async trigger(args) {
      // 檢查參數是否改變 (實際上的檢查邏輯比這個更複雜，也更可靠)
      const argsChanged = Object.is(this.previousArgs, args);
      this.previousArgs = args;

      if (argsChanged) {
        // 如果參數改變了就取消之前的 request
        this.cancelPreviousRequest();
      }

      // 以下任何一個條件符合都會導致 request 被發送：
      // 1. 呼叫 API 的參數改變了
      // 2. cachedData 已經被清除，而且這個 endpoint 目前沒有 request 正在進行
      if (argsChanged || (this.data === undefined && !this.isLoading)) {
        await this.fireRequest(args);
      }
    },
    async fireRequest(args) {
      this.isLoading = true;

      const { data } = await api(args);
      this.data = data;

      this.isLoading = false;
    },
  },
}
```

上面的範例想表達的精神跟 RTK Query 已經很接近了  補上依照指定時間清除 `data` 就大概完成了
針對每個 endpoint 他保證只要傳進來的參數相同  那個 endpoint 就不會發生同時發送多個 request 的情況
不過實際的程式碼肯定比這個更簡潔和完備

然後 `useGetQuery` 的 hook 可以看成是這樣：

```ts
const useGetSomethingQuery = (args) => {
  const { isLoading, data, trigger } = useSelector(
    (state) => state.getSomethingQuery
  );

  useEffect(() => {
    trigger(args);
  }, [args]);

  return {
    isLoading,
    data,
  };
}
```

由於這裡的 store 只有一個  我們的 app 裡面所有的 `const { isLoading, data } = getSomethingQuery()` 都會參考到同一個值  就是 store 裡面的 `getSomethingQuery`
我們在元件中寫下 `const { isLoading, data } = getSomethingQuery()` 的時候  他會用一個 effect 去幫我們呼叫那個 endpoint 的 `trigger` (也許就是 [lazy query](https://redux-toolkit.js.org/rtk-query/api/created-api/hooks#uselazyquery) 的那個 trigger)
假設我們的 app 結構是這樣子：

```jsx
<App>
  <Container>
    <Content />
  </Container>
</App>
```

就算我們在 `App`, `Container` 和 `Content` 三個地方都使用 `useGetSomethingQuery()`  也只會有一個 request 被發送 (`App` 的那個導致的)

所以我認同 ChatGPT 說的
在用 RTK Query 的時候我們不用自己把 response 存進 store 裡面  因為他只會在需要的時候幫我們呼叫 API
所以如果某個元件會需要那個 endpoint 的 response
直接 `useGetSomethingQuery` 下去就可以了  可以放心使用